### PR TITLE
LibLine: Core::EventLoop-ification; a few custom property additions to RPC inspections

### PR DESCRIPTION
**Pull request by alimpfard**
This PR switches LibLine to a Core::Object and its looping to Core::EventLoop.
Individual characters will still block, but that's much better than blocking for an entire line.

Also adds some inspection properties to LibLine, Shell, and ResourceLoader.

I tried making the Property display widget of Inspector a treeview, but due to the lack of a GUI _thing_ processor in my brain, I couldn't do so (In particular, I found `parent_index` annoying to implement, as we don't have a "JsonPath" or anything similar)

### CODE REVIEW

**Comment by alimpfard** [comment]
> Please use variable-like names without spaces in save_to

There are so few of these I couldn't tell the convention. very well.

> why do we need to block on each character

well, if I understand `Core::Notifier` right, it already does a select.
making stdin nonblocking completely breaks everything (dsr in particular), so I opted to leave that for a future PR to deal with.

**Comment by alimpfard** [comment]
...:facepalm:

@bugaevc come up with new pledges at a better time darn it

**Comment by bugaevc** [comment]
> dsr

what's that? 🤔

**Comment by alimpfard** [comment]
Device Status Report
tldr tells you the cursor position in the terminal.
we ask for it once per get_line()

**Comment by alimpfard** [comment]
It always attempted to read 16 bytes at a time, but that `read` will succeed with just 1 byte if you're typing and the event loop will continue

**Comment by bugaevc** [comment]
Well, that makes sense, doesn't it? We `select()` read 1 byte (or up to 16 bytes, if available), process it, `select()` again. `select()` is level-triggered, so it should all work?

**Comment by alimpfard** [comment]
It _does_ work, we just end up waiting for that one byte.

**Comment by bugaevc** [comment]
Waiting as in blocking, on first call? Or wdym?

**Comment by alimpfard** [comment]
we block _once_ per get_line, since it tries to "prime" itself and refresh the display.
otherwise the prompt and stuff would only appear _after_ the user has input something

**Comment by alimpfard** [comment]
Reading what I wrote before, I worded that rather poorly, my bad.

**Comment by bugaevc** [comment]
> otherwise the prompt and stuff would only appear after the user has input something

why is that, why can't it display the prompt without any user input?

**Comment by alimpfard** [comment]
...I'm not sure?
Let me try doing that, and seeing why it breaks.

**Comment by alimpfard** [comment]
Well, that makes sense.
there is no reason to do that at all.

File: Shell/Shell.cpp [hunk 1]
```diff
@@ -1835,3 +1835,26 @@ void Shell::stop_all_jobs()
...
+    Core::Object::save_to(object);
+    object.set("working_directory", cwd);
+    object.set("username", username);
+    object.set("user_home_path", home);
+    object.set("user_id", uid);
+    object.set("directory_stack_size", directory_stack.size());
+    object.set("cd_history_size", cd_history.size());
+
+    // Jobs.
+    object.set("number of jobs", jobs.size());
```
Comment (awesomekling) [inline Shell/Shell.cpp:1860]: How about we put the jobs in a `JsonArray` instead? Then we don't need a separate value for the count.

**Comment by alimpfard** [comment]
:facepalm: I did it again...it's as if every PR is a fix-random-things-around-the-thing-that-the-pr-is-touching for me
